File: Chapter07/src/test/java/com/mastering/spring/springboot/controller/BasicControllerIT.java
package com.mastering.spring.springboot.controller;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.junit4.SpringRunner;

import com.mastering.spring.springboot.Application;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = Application.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class BasicControllerIT {
	private static final String LOCAL_HOST = "http://localhost:";
	@LocalServerPort
	private int port;
	private TestRestTemplate template = new TestRestTemplate();

	@Test
	public void welcome() throws Exception {
		ResponseEntity<String> response = template.getForEntity(createURL("/welcome"), String.class);
		assertThat(response.getBody(), equalTo("Hello World"));
	}

	private String createURL(String uri) {
		return LOCAL_HOST + port + uri;
	}

	@Test
	public void welcomeWithObject() throws Exception {
		ResponseEntity<String> response = template.getForEntity(createURL("/welcome-with-object"), String.class);
		assertThat(response.getBody(), containsString("Hello World"));
	}
	
	 @Test
     public void welcomeWithParameter() throws Exception {
       ResponseEntity<String> response =
         template.getForEntity(
       createURL("/welcome-with-parameter/name/Buddy"), String.class);
       assertThat(response.getBody(),
       containsString("Hello World, Buddy"));
}
}

File: Chapter13/first-web-service-with-kotlin/src/test/kotlin/com/mastering/spring/kotlin/firstwebservicewithkotlin/controller/BasicControllerIT.kt
package com.mastering.spring.kotlin.firstwebservicewithkotlin.controller

import org.hamcrest.Matchers.containsString
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat

import org.junit.Test
import org.junit.runner.RunWith
import org.assertj.core.api.Assertions.assertThat
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.ResponseEntity
import org.springframework.test.context.junit4.SpringRunner
import com.mastering.spring.kotlin.firstwebservicewithkotlin.FirstWebServiceWithKotlinApplication
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.beans.factory.annotation.Autowired
import com.mastering.spring.kotlin.firstwebservicewithkotlin.bean.WelcomeBean

@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class BasicControllerIT {
	@Autowired
	lateinit var restTemplate: TestRestTemplate

	@Test
	fun `GET welcome returns "Hello World"`() {

		// When
		val body = restTemplate.getForObject("/welcome", String::class.java)

		// Then
		assertThat(body).isEqualTo("Hello World")

	}


	@Test
	fun `GET welcome-with-object returns "Hello World"`() {

		// When
		val body = restTemplate.getForObject("/welcome-with-object", WelcomeBean::class.java)

		// Then
		assertThat(body.message, containsString("Hello World"));
	}


	@Test
	fun `GET welcome-with-parameter returns "Hello World"`() {

		// When
		val body = restTemplate.getForObject("/welcome-with-parameter/name/Buddy", WelcomeBean::class.java)

		// Then
		assertThat(body.message, containsString("Hello World, Buddy"));
	}

}